### PR TITLE
(CDAP-15787) 5 Action Plugins

### PR DESCRIPTION
**Pull request by MEseifan**
Implemented FailureCollector into the following action plugins:

- EmailAction
- FileDeleteAction
- FileMoveAction
- SSHAction
- WindowsShareCopy

### CODE REVIEW

File: core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java [hunk 1]
```diff
@@ -107,37 +109,39 @@ public Config() {
...
-    public void validate() {
-      super.validate();
+    public void validate(FailureCollector collector) {
+      super.validate(collector);
 
       if (!containsMacro("username") && (Strings.isNullOrEmpty(username) ^ Strings.isNullOrEmpty(password))) {
-        throw new IllegalArgumentException("You must either set both username and password or " +
-                                             "neither username nor password.");
+        collector.addFailure("You must either set both username and password or neither username nor password.",
+                  "Leave username and password fields empty or provide values for both fields")
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java]: fix indentation to match the rest of the project

Comment (albertshau) [reply core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java]: We don't always need to set a corrective action. When it is basically just restating the error message like here, it's better to leave it out.

File: core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java [hunk 2]
```diff
@@ -107,37 +109,39 @@ public Config() {
...
 
-    public void validate() {
-      super.validate();
+    public void validate(FailureCollector collector) {
+      super.validate(collector);
 
       if (!containsMacro("username") && (Strings.isNullOrEmpty(username) ^ Strings.isNullOrEmpty(password))) {
-        throw new IllegalArgumentException("You must either set both username and password or " +
-                                             "neither username nor password.");
+        collector.addFailure("You must either set both username and password or neither username nor password.",
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java]: this was an existing problem, but most places we don't use 'You'. Change this to:

Both username and password must be given, or neither of them must be given.

File: core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java [hunk 3]
```diff
@@ -107,37 +109,39 @@ public Config() {
...
+                  "Leave username and password fields empty or provide values for both fields")
+                  .withConfigProperty("username").withConfigProperty("password");
       }
 
       if (!containsMacro("sender")) {
         try {
           InternetAddress[] addresses = InternetAddress.parse(sender);
           if (addresses.length == 0) {
-            throw new IllegalArgumentException("Must specify a sender email address.");
+            collector.addFailure("Sender email was not specified.", "Provide a sender email address.")
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java]: When we complain that some required property is missing, we can just use a single error message with no corrective action.

"Sender email must be specified."

should be good enough

File: core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java [hunk 4]
```diff
@@ -107,37 +109,39 @@ public Config() {
...
+                    .withConfigProperty("sender");
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java]: create `private static final String` constants for these

File: core-plugins/src/main/java/io/cdap/plugin/batch/action/FileDeleteAction.java [hunk 5]
```diff
@@ -126,13 +131,13 @@ public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
...
-    public void validate() {
+    public void validate(FailureCollector collector) {
       if (!containsMacro("fileRegex") && fileRegex != null) {
         try {
           Pattern.compile(fileRegex);
         } catch (Exception e) {
-          throw new IllegalArgumentException(String.format("File regex %s is invalid: %s",
-                                                           fileRegex, e.getMessage()), e);
+          collector.addFailure(String.format("File regex %s is invalid: %s", fileRegex, e.getMessage()),
+                  "Provide valid regex.").withConfigProperty("fileRegex");
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/action/FileDeleteAction.java]: create constants for these, similar comment elsewhere

File: core-plugins/src/main/java/io/cdap/plugin/batch/action/WindowsShareCopy.java [hunk 6]
```diff
@@ -58,225 +60,252 @@
...
-                return copyFileToHDFS(hdfs, smbDirectory + file, hdfsDir, auth);
-              } else {
-                return copyFileToHDFS(hdfs, smbDirectory + "/" + file, hdfsDir, auth);
-              }
-            } catch (Exception e) {
-              LOG.warn("Exception while copying the file {}", file, e);
-              return null;
+    @Override
+    public void run(ActionContext context) throws Exception {
+        config.numThreads = (config.numThreads == null || config.numThreads < MIN_NUM_THREADS) ? MIN_NUM_THREADS :
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/action/WindowsShareCopy.java]: indentation

File: hydrator-common/src/main/java/io/cdap/plugin/common/batch/action/ConditionConfig.java [hunk 7]
```diff
@@ -44,7 +45,7 @@ public ConditionConfig(String runCondition) {
   }
 
   @SuppressWarnings("ConstantConditions")
-  public void validate() {
+  public void validate(FailureCollector collector) {
```
Comment (albertshau) [inline hydrator-common/src/main/java/io/cdap/plugin/common/batch/action/ConditionConfig.java:65]: this looks incomplete

File: core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java (hunk 1 above)
Comment (MEseifan) [reply core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java]: Removed corrective action and reformatted file

File: core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java (hunk 3 above)
Comment (MEseifan) [reply core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java]: Removed corrective action for all missing property errors

File: core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java (hunk 2 above)
Comment (MEseifan) [reply core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java]: Done

File: core-plugins/src/main/java/io/cdap/plugin/batch/action/WindowsShareCopy.java (hunk 6 above)
Comment (MEseifan) [reply core-plugins/src/main/java/io/cdap/plugin/batch/action/WindowsShareCopy.java]: Done

File: core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java (hunk 4 above)
Comment (MEseifan) [reply core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java]: Done

File: core-plugins/src/main/java/io/cdap/plugin/batch/action/FileDeleteAction.java (hunk 5 above)
Comment (MEseifan) [reply core-plugins/src/main/java/io/cdap/plugin/batch/action/FileDeleteAction.java]: Done

File: hydrator-common/src/main/java/io/cdap/plugin/common/batch/action/ConditionConfig.java (hunk 7 above)
Comment (MEseifan) [reply hydrator-common/src/main/java/io/cdap/plugin/common/batch/action/ConditionConfig.java:65]: Updated it to use the collector

File: core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java [hunk 8]
```diff
@@ -107,37 +116,39 @@ public Config() {
...
-      super.validate();
+    public void validate(FailureCollector collector) {
+      super.validate(collector);
 
-      if (!containsMacro("username") && (Strings.isNullOrEmpty(username) ^ Strings.isNullOrEmpty(password))) {
-        throw new IllegalArgumentException("You must either set both username and password or " +
-                                             "neither username nor password.");
+      if (!containsMacro(USERNAME) && (Strings.isNullOrEmpty(username) ^ Strings.isNullOrEmpty(password))) {
+        collector.addFailure("Both username and password must be given, or neither of them must be given.",
+                            "Leave username and password fields empty or provide values for both fields")
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java]: indentation off by 1 space

File: core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java [hunk 9]
```diff
@@ -107,37 +116,39 @@ public Config() {
...
+                .withConfigProperty(USERNAME).withConfigProperty(PASSWORD);
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java]: method calls should be indented 2 spaces:

```
collector.addFailure(...)
  .withConfigProperty(...)
```

File: core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java [hunk 10]
```diff
@@ -107,37 +116,39 @@ public Config() {
...
       }
 
-      if (!containsMacro("sender")) {
+      if (!containsMacro(SENDER)) {
         try {
           InternetAddress[] addresses = InternetAddress.parse(sender);
           if (addresses.length == 0) {
-            throw new IllegalArgumentException("Must specify a sender email address.");
+            collector.addFailure("Sender email was not specified.", "")
+                    .withConfigProperty(SENDER);
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java]: indentation

File: core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java [hunk 11]
```diff
@@ -107,37 +116,39 @@ public Config() {
...
           }
           if (addresses.length > 1) {
-            throw new IllegalArgumentException(String.format(
-              "%s is an invalid sender email address. Only one sender is supported.", sender));
+            collector.addFailure(String.format(
+                    "%s is an invalid sender email address. Only one sender is supported.", sender),
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java]: indentation

File: core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java [hunk 12]
```diff
@@ -107,37 +116,39 @@ public Config() {
...
+                    "Only specify one sender email address.").withConfigProperty(SENDER);
           }
         } catch (AddressException e) {
-          throw new IllegalArgumentException(
-            String.format("%s is an invalid sender email address. Reason: %s", sender, e.getMessage()));
+          collector.addFailure(String.format("%s is an invalid sender email address. Reason: %s", sender,
+                                e.getMessage()), "").withConfigProperty(SENDER);
```
Comment (albertshau) [inline core-plugins/src/main/java/io/cdap/plugin/batch/action/EmailAction.java]:
```
.addFailure(String.format("...",
                          e.getMessage()),
            null)
  .withConfigProperty()
```

**Review by albertshau — COMMENTED** [review]
logic looks fine but indentation is off, please fix across all PRs

**Review by albertshau — APPROVED** [review]
lgtm, please squash commits before merging

**Comment by albertshau** [comment]
https://builds.cask.co/browse/HYP-BAD590

File: hydrator-common/src/main/java/io/cdap/plugin/common/batch/action/ConditionConfig.java [hunk 13]
```diff
@@ -44,13 +45,13 @@ public ConditionConfig(String runCondition) {
...
```
Comment (CuriousVini) [inline hydrator-common/src/main/java/io/cdap/plugin/common/batch/action/ConditionConfig.java:65]: This change should be backwards compatible. We should deprecate this method and add a new method with failure collector.

Comment (CuriousVini) [reply hydrator-common/src/main/java/io/cdap/plugin/common/batch/action/ConditionConfig.java:65]: Fixed.

**Review by albertshau — APPROVED** [review]

**Comment by CuriousVini** [comment]
`SSHActionTest` test passed locally. Squashing and merging.

**Pull request merged**